### ISSUE TITLE
[Docker Desktop] 4.9.x, 4.10.x, 4.11.x, 4.12.x, 4.13.x, 4.14.x and 4.15.x don't install properly on Fedora

### ISSUE THREAD

**Issue by amharris** [docker/desktop-linux#36]
**Tell us about your request**
Docker Desktop 4.9.0 doesn't install properly and causes problems with filesystem permissions in containerised applications.

**Which service(s) is this request for?**
Docker Desktop (and any parts of Docker Engine, if also affected)

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
When uninstalling Docker Desktop and installing the updated release—4.9.0 (80466)—per the documentation, it doesn't install properly.

Containerised projects of mine also have storage- and/or ownership-related issues with this version, once installed.

**Are you currently working around the issue?**
Falling back to 4.8.2 (79419) works for now.

**Additional context**
Running on: Fedora (36) Workstation

The output of the installation process for 4.9.0 (80466), with the error, is as follows:

```
Running transaction
  Preparing        :                                                        1/1 
  Installing       : docker-desktop-4.9.0-80466.x86_64                      1/1 
  Running scriptlet: docker-desktop-4.9.0-80466.x86_64                      1/1 
Enabling use of privileged ports by Docker Desktop
kubernetes.docker.internal already in /etc/hosts
Reloading systemd daemon for logged in users
Done reloading systemd daemon for logged in users

ln: failed to create symbolic link '/usr/local/bin/com.docker.cli': File exists
Enabling use of privileged ports by Docker Desktop

  Verifying        : docker-desktop-4.9.0-80466.x86_64                      1/1 

Installed:
  docker-desktop-4.9.0-80466.x86_64                                             
```

Please note that Docker Engine is also installed on the machine, given the referenced binary above, (in case that's why this conflict has occurred; though Docker Engine and Docker Desktop are supposed to be able to coexist on the same host).

I should also note that the installation runs extremely quickly; too quick, in fact.

When it comes to using Sail—Laravel's Docker environment—an existing installation of a project would be met with `Unable to create lockable file: /var/www/html/storage/framework/cache/data/...` when trying to register a user or login, and a new installation of a project would be met with `the stream or file "/var/www/html/storage/logs/laravel.log" could not be opened in append mode: failed to open stream: permission denied "sail"`.

---

I thought it would also be of value to attempt an upgrade, rather than uninstalling the previous version (which is the intended approach as outlined in the documentation), and here's the result:

```
Running transaction
  Preparing        :                                                        1/1 
  Upgrading        : docker-desktop-4.9.0-80466.x86_64                      1/2 
  Running scriptlet: docker-desktop-4.9.0-80466.x86_64                      1/2 
  Running scriptlet: docker-desktop-4.8.2-79419.x86_64                      2/2 
Stopping docker-desktop for logged in users
Warning: The unit file, source configuration file or drop-ins of docker-desktop.service changed on disk. Run 'systemctl --user daemon-reload' to reload units.
Done stopping docker-desktop for logged in users

  Cleanup          : docker-desktop-4.8.2-79419.x86_64                      2/2 
  Running scriptlet: docker-desktop-4.8.2-79419.x86_64                      2/2 
Reloading systemd daemon for logged in users
Done reloading systemd daemon for logged in users

  Running scriptlet: docker-desktop-4.9.0-80466.x86_64                      2/2 
Enabling use of privileged ports by Docker Desktop

  Running scriptlet: docker-desktop-4.8.2-79419.x86_64                      2/2 
  Verifying        : docker-desktop-4.9.0-80466.x86_64                      1/2 
  Verifying        : docker-desktop-4.8.2-79419.x86_64                      2/2 

Upgraded:
  docker-desktop-4.9.0-80466.x86_64                                             
```

There's no clear symbolic link error but, after starting it up, it still remains to be an issue all the same.

---

Finally, uninstalling and going back to 4.8.2 (79419), this is the install output:

```
Running transaction
  Preparing        :                                                        1/1 
  Installing       : docker-desktop-4.8.2-79419.x86_64                      1/1 
  Running scriptlet: docker-desktop-4.8.2-79419.x86_64                      1/1 
Enabling use of privileged ports by Docker Desktop
kubernetes.docker.internal already in /etc/hosts
Reloading systemd daemon for logged in users
Done reloading systemd daemon for logged in users

  Verifying        : docker-desktop-4.8.2-79419.x86_64                      1/1 

Installed:
  docker-desktop-4.8.2-79419.x86_64                                             
```

The output is as expected, just like 4.8.1 (78998) and 4.8.0 (78933), and spinning it up and running an existing project or new installation of a containerised project works just as expected.

**Comment by stephen-turner** [comment]
This looks like a bug, not a feature request, so moving to desktop-linux repository.

**Comment by amharris** [comment]
> This looks like a bug, not a feature request, so moving to desktop-linux repository.

Apologies for that, Stephen.  I saw many other bugs being raised here, and was driven towards this repo via the docs, so didn't think more of it than that.

I'll note that for the future. 👍

**Comment by p1-0tr** [comment]
@amharris - sorry for the long wait. Does the issue still occur on the most recent releases of Docker Desktop? Also could you share a little more information about how you run your project?

**Comment by amharris** [comment]
Hi, @p1-0tr.  Not a problem; and glad to hear back.

I've given 4.11.x a whirl and it, too, remains an issue:
```
Running transaction
  Preparing        :                                                        1/1 
  Installing       : docker-desktop-4.11.1-84025.x86_64                     1/1 
  Running scriptlet: docker-desktop-4.11.1-84025.x86_64                     1/1 
Enabling use of privileged ports by Docker Desktop
kubernetes.docker.internal already in /etc/hosts
Reloading systemd daemon for logged in users
Done reloading systemd daemon for logged in users

ln: failed to create symbolic link '/usr/local/bin/com.docker.cli': File exists
Enabling use of privileged ports by Docker Desktop

  Verifying        : docker-desktop-4.11.1-84025.x86_64                     1/1 

Installed:
  docker-desktop-4.11.1-84025.x86_64                                            
```

The moment I revert back to 4.8.2, it's fine once again, being the last working version for me.  It's probably worth me reiterating that 4.8.2 takes considerably longer to install (in a good way), whereas 4.9.x, 4.10.x, and 4.11.x all 'complete' far too quickly, (whilst throwing the error; and, whilst Docker Desktop itself starts, the underlying components are evidently not in a fully working state, so projects won't spin up).

#### Re. your questions
This will occur before I run my project, so I don't think it should have too much bearing as to the projects being run, but so you've got a control to work with, here's how you can reproduce it (using Fedora 36 Workstation) and the framework of the project in use:

A standard, fresh Laravel installation with Sail (a project from its own ecosystem that runs and manages a set of Docker scripts) will do the trick; either by spinning it up as a Sail project with [the documentation's bash script](https://laravel.com/docs/9.x#getting-started-on-linux) or [beginning fresh via Composer](https://laravel.com/docs/9.x#your-first-laravel-project) and [adding in Sail](https://laravel.com/docs/9.x/sail#installing-sail-into-existing-applications), and [spinning it up](https://laravel.com/docs/9.x/sail#starting-and-stopping-sail) (with either method, that should result in near-identical project structures).

The project is only being run within home, under a subdirectory for projects, (like you would have for documents, music, etc.)


Many thanks, and hopefully this'll be addressed soon.

**Comment by amharris** [comment]
Hi, @p1-0tr,

Just to update you; this remains the case with the recent release, 4.12.

**Comment by amharris** [comment]
Any news on this, @p1-0tr?

Hopefully this is being investigated?

**Comment by p1-0tr** [comment]
hi @amharris, yes. We are in the process of changing the way file sharing between the host and the DD VM is implemented (specifically in regard to how we deal with file permissions), which seems to be the root cause of your problem. I'll put together a dev build for you to try tomorrow, or failing that, on Monday.

**Comment by amharris** [comment]
Hi @p1-0tr,

Thank you for getting back to me.
That's fantastic news.  I'll eagerly be awaiting that dev build.

I'll let you know as soon as I've tested it (should be on the day, whether tomorrow or Monday) and let you know the results.

**Comment by p1-0tr** [comment]
hi @amharris, I've generated a dev build with the file sharing changes in it

* https://desktop-stage.docker.com/linux/main/amd64/94103/docker-desktop-4.16.0-x86_64.rpm
* https://desktop-stage.docker.com/linux/main/amd64/94103/docker-desktop-4.16.0-amd64.deb

I've ran the example laravel app with this build successfully, but please note it is still in development. That said I will greatly appreciate any feedback you may have.

**Comment by amharris** [comment]
Hi, @p1-0tr 

Thanks for that.

The good news is that Laravel applications using Sail now work.

The bad news is that symbolic link error still occurs upon installation (with a suspiciously quick installation time).
I did previously try wiping all config and cache data for Docker Desktop, along with removing all CLI packages via the CE repository, and letting the installation of Docker Desktop handle that, but the error would always appear.  The last version to not display that with the RPM build (at least on Fedora Workstation) also being 4.8.x.

```
Last metadata expiration check: 0:00:01 ago on Fri 16 Dec 2022 15:23:54 GMT.
Dependencies resolved.
===============================================================================================================================
 Package                          Architecture             Version                        Repository                      Size
===============================================================================================================================
Installing:
 docker-desktop                   x86_64                   4.16.0-94103                   @commandline                   503 M

Transaction Summary
===============================================================================================================================
Install  1 Package

Total size: 503 M
Installed size: 1.5 G
Is this ok [y/N]: y
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                       1/1 
  Installing       : docker-desktop-4.16.0-94103.x86_64                                                                    1/1 
  Running scriptlet: docker-desktop-4.16.0-94103.x86_64                                                                    1/1 
Enabling use of privileged ports by Docker Desktop
kubernetes.docker.internal already in /etc/hosts
Reloading systemd daemon for logged in users
Done reloading systemd daemon for logged in users

ln: failed to create symbolic link '/usr/local/bin/com.docker.cli': File exists
Enabling use of privileged ports by Docker Desktop

  Verifying        : docker-desktop-4.16.0-94103.x86_64                                                                    1/1 

Installed:
  docker-desktop-4.16.0-94103.x86_64                                                                                           

Complete!
```

I haven't tried the dev `.deb` build above, as I don't have an installation on-hand right now but, as I may have mentioned before, that one never broke in my experience; at least with trying it out on Ubuntu, it seemed to be fine (neither that error, nor issues with Laravel Sail applications starting).  Though I only did that out of curiosity, to see if the problem was consistent across packaged builds or not.
Though if you'd like me to test that too, to ensure no regressions, let me know.


Most importantly, though, good news on the applications now running, so thanks for compiling that build for me, and many thanks to the team for sorting that out.

**Comment by p1-0tr** [comment]
@amharris thanks for the feedback :)

Hmm, it's a bit weird that the permissions problem did not occur on Ubuntu. I'll have to look into that.

The symlink error being printed should not matter (it does not prevent installation/upgrade from making progress). I have done a fix for it, but I have to have another look at the upgrade path (installing over an existing `docker-desktop`) package.

**Comment by amharris** [comment]
Hi @p1-0tr 

No problem.

I think at the time, when it initially became a problem with 4.9.x, given it was correlated with the larger issue, I presumed that it was related—especially given the speed at which the installation would take place, my suspicion was that some core parts weren't being installed as a consequence of the failed symlink.

What I noticed, originally, was that the binary isn't there when the symlink fails, but with 4.8.x and prior, when it was fine, it was.  I _may_ be incorrect here, as I'm only responding via mobile and don't have a machine on hand, in order to verify, so my own memory may be suffering from a little 'human bit-rot'.

But yes, an upgrade path from 4.8.x to 4.16.x (dev build) is fine, at least.
I have yet to test if there are different conflicts/outcomes depending on whether parts of Docker Engine are installed differently, or have never been set up (effecting the permission), before a clean installation of Docker Desktop 4.16.x (dev).  If you'd like me to test on a clean installation, directly to your dev build, let me know.

As for your ending comment; yes, it would be nice to know if that could be worked on.  I only followed the approach as officially outlined in the documentation, of removing the existing one and installing thereafter.
Though the ideal would be for packaged distribution, as with Docker Engine.  If you don't mind me asking, what's the current blocker (policies or technicalities) on that not being the case at present, given it is and has been for a long time with Docker Engine?

**Comment by p1-0tr** [comment]
hi @amharris, thanks for the extra information. If 4.8.x -> (dev build) update works well, it should also be fine for more recent versions (the cli executable has not changed in a fair while). I will follow up when we release the permissions fixes (most likely the February release), though.

As to getting the package into our package repo, it is on our radar (unfortunately I can't put a date on it right now).

**Comment by amharris** [comment]
Hi, @p1-0tr,

No problem.  I'm more than happy to provide information that furthers helping you to help me (and any others with the same issue).

That's fine and good to hear; I'm happy to roll with the dev build you provided for now.  Hopefully the release around February time will include a fix for symlink error for the CLI executable (and it seemingly not being properly installed, (perhaps the reason behind the unusually fast installation times, compared with 4.8.x)).

I don't know what your current testing and deployment infrastructure looks like, but with Linux support (for Docker Desktop) still being fairly fresh, I can understand if automated testing around the builds is fairly minimal at the moment.
If fairly simple at the moment, I think it'd be worth putting in the time to test the installation for each release and capturing the log output—at least for the officially-supported distributions (Debian, Ubuntu, Fedora, etc.) with their respectively built packages which you provide.
Ideally automated, perhaps by capturing the logs and `grep`ing for any current or previously-known errors, (along with some general key words); or even a quick manual test and some eyes going over it (with the latest stable version of the distribution, up-to-date, or, in the case of Ubuntu, both the latest stable and latest LTS) to watch for regressions or blocker-level problems.

No worries with respect to putting a date on getting it into the package repository.
I would assume that having some more robust testing around each build is probably a prerequisite in your mind, anyway, as you'll no doubt want fairly smooth sailing when users run `dnf update`/`apt update`.
Though I'm glad to hear it's somewhere on the road-map. 👍🏻

**Comment by amharris** [comment]
Hi, @p1-0tr,

I just wanted to reach out to let you know I saw the formal release of 4.16.0 and 4.16.1—a fair bit earlier than you were expecting, it seems.

Anyway, the good news is that the fixes you included in the dev build still seem to remain; and, on top of that, the symlink error has disappeared, with a clean install taking place.  I've already performed this on a couple of systems with complete success.

I only noticed the update last night, so I skipped from the dev build to 4.16.1.  I don't know if both were included in 4.16.0.
However, judging by the change-logs, I can't see that either are mentioned under the Linux section, so I thought it best to let you know, so you can update those.

A big thank-you to you and the team for sorting both of these issues out.
I'll leave it to you to update the change-logs and then this issue can be closed; marked as resolved 🙂

**Comment by p1-0tr** [comment]
Hi @amharris, The file-sharing changes from the dev build have not been merged yet, so I am a bit confused :D The symlink fix should have been fixed, so I am glad that it is indeed so. Out of curiosity - did you uninstall and/or stop Docker Desktop before updating? If not, could you restart Docker Desktop manually, to make sure that it's not just the dev build running (which would point to our systemd integration being flaky)?

**Comment by amharris** [comment]
Hi @p1-0tr,

Thanks for getting back to me.

I can say with absolute certainty that, on both systems, the dev build was stopped, uninstalled, and then 4.16.1 was installed after that.

Whether running immediately afterwards or rebooting first (I tried both; one on each system), it worked.

Is it possible that the installation of the dev build had some kind of lasting effect?
It's very bizarre, if you're certain this shouldn't have worked, but I'll happily install every version from 4.8.x, including the dev build of 4.16 that you provided me, and log the result.  Though I believe I've done that with this GitHub issue.

Perhaps the symlink issue being addressed did fix the problem, but migrating from 4.8.x to the dev build was enough to get around it in part (with the file-sharing changes).  So, is it possible that each on their own would've been enough to run Laravel Sail-based projects, but both together are the complete solution?
I suppose I could try uninstalling 4.16.1, completely wiping all of the associated dot-files and configs, and then performing a clean installation.  This will mean wiping the project, so I'm happy to perform this on my personal machine (rather than my work one).  It'd be interesting to see the result of that.

Anyway, I'll let you take a closer look; and if I can do any more digging, let me know.
